src/libcore/convert/num.rs

```diff
@@ -445,3 +445,38 @@ nzint_impl_from! { NonZeroU16, NonZeroI128, #[stable(feature = "nz_int_conv", si
 nzint_impl_from! { NonZeroU32, NonZeroI64, #[stable(feature = "nz_int_conv", since = "1.41.0")] }
 nzint_impl_from! { NonZeroU32, NonZeroI128, #[stable(feature = "nz_int_conv", since = "1.41.0")] }
 nzint_impl_from! { NonZeroU64, NonZeroI128, #[stable(feature = "nz_int_conv", since = "1.41.0")] }
+
+macro_rules! nzint_impl_try_from {
+    (@PER_ITEM  $NonZeroType: ident; $SameButZeroType: ident; $PossiblyZeroType: ty; #[$attr:meta]; $doc: expr) => {
+                #[$attr]
+                #[doc = $doc]
+                impl TryFrom<$PossiblyZeroType> for $NonZeroType {
+                    type Error = TryFromIntError;
+                    #[inline]
+                    fn try_from(value: $PossiblyZeroType) -> Result<Self,Self::Error> {
+                        <$SameButZeroType as TryFrom<$PossiblyZeroType>>::try_from(value)
```

Comment:
This could be written without going through the Iterator interface using and_then:

```rust
<$SameButZeroType as TryFrom<$PossiblyZeroType>>::try_from(value)
    .ok().and_then($NonZeroType::new).ok_or(TryFromIntError(()))
```